src/main/kotlin/org/rust/cargo/toolchain/Cargo.kt

```diff
@@ -94,12 +100,49 @@ class Cargo(private val cargoExecutable: Path) {
         owner: Project,
         projectDirectory: Path,
         listener: ProcessListener? = null
-    ): String {
+    ): CargoWorkspaceData {
+        val rawData = fetchMetadata(owner, projectDirectory, listener)
+        val buildPlan = fetchBuildPlan(owner, projectDirectory, listener)
+        return CargoMetadata.clean(rawData, buildPlan)
+    }
+
+    @Throws(ExecutionException::class)
+    private fun fetchMetadata(
+        owner: Project,
+        projectDirectory: Path,
+        listener: ProcessListener?
+    ): CargoMetadata.Project {
         val additionalArgs = mutableListOf("--verbose", "--format-version", "1", "--all-features")
-        return CargoCommandLine("metadata", projectDirectory, additionalArgs)
+        val json = CargoCommandLine("metadata", projectDirectory, additionalArgs)
             .execute(owner, listener = listener)
             .stdout
             .dropWhile { it != '{' }
+        return try {
+            Gson().fromJson(json, CargoMetadata.Project::class.java)
+        } catch (e: JsonSyntaxException) {
+            throw ExecutionException(e)
+        }
+    }
+
+    private fun fetchBuildPlan(
+        owner: Project,
+        projectDirectory: Path,
+        listener: ProcessListener?
+    ): CargoBuildPlan? {
+        val additionalArgs = mutableListOf("-Z", "unstable-options", "--all-targets", "--build-plan")
+        // Hack to make cargo think that current channel is nightly because we need unstable `--build-plan` option here
+        val envs = EnvironmentVariablesData.create(mapOf(
+            RUSTC_BOOTSTRAP to "1"
```

Comment:
😨

Reply:
This means that this functionality can break at any moment.

Reply:
We are closely monitoring changes in the rust toolchain and ready for breaking changes in the build-plan or other nightly features that we use. We'd want a more explicit way to unlock all nightly command-line options on the stable toolchain but now there is only "RUSTC_BOOTSTRAP" magic.